File: lib/src/main/java/com/fidelyo/messenger/Subscriber.java
package com.fidelyo.messenger;

import android.os.Handler;

import java.util.UUID;

/**
 * Created by bishoy on 2/22/18.
 */

public class Subscriber<T> {

    private final UUID uuid;
    private final String key;
    private final Handler handler;
    private Callback<T> callback = t -> {
    };

    public Subscriber(String key, Handler handler) {
        this.key = key;
        this.handler = handler;

        uuid = UUID.randomUUID();

    }

    public String getKey() {
        return key;
    }

    public void handle(Callback<T> callback) {
        this.callback = callback;
    }

    public void publish(Object o) {
        handler.post(() -> callback.onMessage((T) o));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Subscriber<?> that = (Subscriber<?>) o;

        return uuid.equals(that.uuid);
    }

    @Override
    public int hashCode() {
        return uuid.hashCode();
    }
}

File: lib/src/main/java/com/fidelyo/messenger/Subscriber.kt
package com.fidelyo.messenger

import android.os.Handler

/**
 * Created by bishoy on 7/21/17.
 */
class Subscriber<T>(internal var key: String,
                    internal var handler: Handler) {

    internal var id: Long? = null
    private var callback: Callback<T>? = null

    fun handle(handler: (T) -> Unit) {
        callback = object : Callback<T> {
            override fun onMessage(t: T) {
                handler(t)
            }
        }
    }

    internal fun publish(obj: Any) {
        handler.post { callback?.onMessage(obj as T) }
    }

}


